pkg/processor/limiter/processor.go

```diff
@@ -0,0 +1,146 @@
+	logger  *zap.Logger
+	storage *storage.Storage
+}
+
+func newLimiterProcessor(logger *zap.Logger) (*limiterProcessor, error) {
+	return &limiterProcessor{
+		logger:  logger,
+		storage: env.G().Storage(),
+	}, nil
+}
+
+func (lp *limiterProcessor) ConsumeTraces(ctx context.Context, td ptrace.Traces) (ptrace.Traces, error) {
+	err := lp.consume(
+		ctx,
+		entity.SignalTraces,
+		entity.NewLimitValue(
+			td.SpanCount(),
+			(&ptrace.ProtoMarshaler{}).TracesSize(td),
+		),
+	)
+
+	if err != nil {
+		return td, err
+	}
+
+	return td, nil
+}
+
+func (lp *limiterProcessor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	err := lp.consume(
+		ctx,
+		entity.SignalMetrics,
+		entity.NewLimitValue(
+			md.MetricCount(),
+			(&pmetric.ProtoMarshaler{}).MetricsSize(md),
+		),
+	)
+
+	if err != nil {
+		return md, err
+	}
+
+	return md, nil
+}
+
+func (lp *limiterProcessor) ConsumeLogs(ctx context.Context, ld plog.Logs) (plog.Logs, error) {
+	err := lp.consume(
+		ctx,
+		entity.SignalLogs,
+		entity.NewLimitValue(
+			ld.LogRecordCount(),
+			(&plog.ProtoMarshaler{}).LogsSize(ld),
+		),
+	)
+
+	if err != nil {
+		return ld, err
+	}
+
+	return ld, nil
+}
+
+func (lp *limiterProcessor) consume(ctx context.Context, signal entity.Signal, limitValue entity.LimitValue) error {
+	// Retrieve credential id from ctx
+	keyId, ok := entity.KeyIdFromCtx(ctx)
+	if !ok {
+		return nil
+	}
+
+	// Retrieve the limit from the backing store
+	limit, err := lp.storage.DAO.Limits().SelectByKeyAndSignal(ctx, keyId, signal)
+	if err != nil {
+		t, _, _ := errors.Unwrapb(err)
+		if t == errors.TypeNotFound {
+			return nil
+		}
+
+		return err
+	}
+
+	// Set the current time here for rate limiting
+	currTime := time.Now()
+
+	// Iterate over all periods and check whether limit has exceeded or not
+	for period, maxLimitValue := range limit.Config.Map() {
+		currLimitValue, err := lp.storage.DAO.LimitMetrics().
+			SelectValueByPeriodAndSignalAndPeriodAtAndKeyId(
+				ctx,
+				period,
+				signal,
+				entity.PeriodAt(currTime)[period],
+				keyId,
+			)
+		if err != nil {
+			t, _, _ := errors.Unwrapb(err)
+			if t != errors.TypeNotFound {
+				return err
+			}
+			// If the current limit metric for the specified period was not found, initialize to zero
+			currLimitValue = entity.NewLimitValue(0, 0)
+		}
+
+		// Return error on the period that led to the limit
+		if maxLimitValue.Subtract(currLimitValue).Subtract(limitValue).LessThan(0) {
+			return status.Error(codes.ResourceExhausted, "limit exceeded")
+		}
+	}
+
+	// Otherwise increment the limit metrics and only create entries for limits specified in the config map
+	var limitMetrics []*entity.LimitMetric
+	for period := range limit.Config.Map() {
+		limitMetrics = append(limitMetrics, entity.NewLimitMetric(
+			period,
+			limitValue,
+			signal,
+			entity.PeriodAt(currTime)[period],
+			keyId,
+		))
+	}
+
+	err = lp.storage.DAO.LimitMetrics().Insert(ctx, limitMetrics)
```

Comment:
The point we were trying to make during the discussion was that the processor would maintain a temporary local state of aggregation for a minute or so and flush them in intervals. The current limits also do not need to be fetched for each consume request.

Reply:
The aggregation has to be done on `period, signal, period_at, key_id`. Unless we use a consistent hash load balancer in front to ensure all requests with the same `key_id` hit the same instance, we would just be delaying the insertion of all independent metrics.

Reply:
It doesn't have to be consistent. The trade-off is that the local state is behind the actual value and allows ingesting a little bit more than actually permitted. And with a reasonable flush interval, that over commit of data won't make much difference.

Reply:
I don't follow. I am saying there is no need to have a consistent hash load balancer because the count/sizes can lag for the duration of the flush interval. The component maintains a local aggregation state per each unique period, signal, key_id etc.. And update the count/size once every flush interval.

For example, take key `k1` and four collectors `c1-4`. For the sake of argument, say the `consume` was called 10000 times for with `k1` and counts are `c1-2000, c2-3000, c3-4000, c5-1000`. The collector does the local aggregation for a minute and at the end of the minute interval it updates the counts in the persistent store.